### PR TITLE
expression: rewrite builtin function: UNIX_TIMESTAMP

### PR DESCRIPTION
**Pull request by breeswish**
To https://github.com/pingcap/tidb/issues/4080

Some test cases are commented out due to:
- #4231 
- #4232 
- #4233 
- #4296

### CODE REVIEW

File: expression/builtin_time.go [hunk 1]
```diff
@@ -1908,64 +1910,127 @@ func (c *unixTimestampFunctionClass) getFunction(args []Expression, ctx context.
...
+	var retFLen, retDecimal int
 
-type builtinUnixTimestampSig struct {
-	baseBuiltinFunc
-}
+	if len(args) == 0 {
+		retTp, retDecimal = tpInt, 0
+	} else {
+		argTps = []evalTp{tpTime}
+		fieldType := args[0].GetType()
```
Comment (zz-jason) [inline expression/builtin_time.go]: s/fieldType/argType/

File: expression/builtin_time.go [hunk 2]
```diff
@@ -1908,64 +1910,127 @@ func (c *unixTimestampFunctionClass) getFunction(args []Expression, ctx context.
...
+		tp := fieldTp2EvalTp(fieldType)
```
Comment (zz-jason) [inline expression/builtin_time.go]: s/tp/argTp/

File: expression/builtin_time.go [hunk 3]
```diff
@@ -1908,64 +1910,127 @@ func (c *unixTimestampFunctionClass) getFunction(args []Expression, ctx context.
...
+		} else {
+			retTp = tpDecimal
+		}
+	}
+	if retTp == tpInt {
+		retFLen = 11
+	} else if retTp == tpDecimal {
+		retFLen = 12 + retDecimal
+	} else {
+		panic("Unexpected retTp")
```
Comment (zz-jason) [inline expression/builtin_time.go]: we should throw an error instead of panic

File: expression/builtin_time.go [hunk 4]
```diff
@@ -1908,64 +1910,127 @@ func (c *unixTimestampFunctionClass) getFunction(args []Expression, ctx context.
...
-		now := time.Now().Unix()
-		d.SetInt64(now)
-		return
+		sig = &builtinUnixTimestampCurrentSig{baseIntBuiltinFunc{bf}}
+	} else if retTp == tpInt {
+		sig = &builtinUnixTimestampIntSig{baseIntBuiltinFunc{bf}}
+	} else if retTp == tpDecimal {
+		sig = &builtinUnixTimestampDecSig{baseDecimalBuiltinFunc{bf}}
+	} else {
+		panic("Unexpected retTp")
```
Comment (zz-jason) [inline expression/builtin_time.go]: ditto

File: expression/builtin_time.go [hunk 5]
```diff
@@ -1908,64 +1910,127 @@ func (c *unixTimestampFunctionClass) getFunction(args []Expression, ctx context.
...
+}
 
-	t1, err = t.Time.GoTime(getTimeZone(b.ctx))
+// evalInt evals a UNIX_TIMESTAMP(time).
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_unix-timestamp
+func (b *builtinUnixTimestampIntSig) evalInt(row []types.Datum) (int64, bool, error) {
+	val, isNull, err := b.args[0].EvalTime(row, b.getCtx().GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		// Return 0 for invalid date time.
+		return 0, isNull, nil
```
Comment (zz-jason) [inline expression/builtin_time.go]: should we handle the invalid time error here ?

File: expression/builtin_time.go [hunk 6]
```diff
@@ -1908,64 +1910,127 @@ func (c *unixTimestampFunctionClass) getFunction(args []Expression, ctx context.
...
+	baseDecimalBuiltinFunc
+}
+
+// evalDecimal evals a UNIX_TIMESTAMP(time).
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_unix-timestamp
+func (b *builtinUnixTimestampDecSig) evalDecimal(row []types.Datum) (*types.MyDecimal, bool, error) {
+	val, isNull, err := b.args[0].EvalTime(row, b.getCtx().GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		// Return 0 for invalid date time.
+		return new(types.MyDecimal), isNull, nil
```
Comment (zz-jason) [inline expression/builtin_time.go]: ditto

File: expression/builtin_time.go (hunk 5 above)
Comment (breeswish) [inline expression/builtin_time.go]: we should return zero in this function.

**Comment by breeswish** [comment]
@zz-jason Done, PTAL

**Comment by zz-jason** [comment]
@breeswish plz resolve conflicts, @XuHuaiyu PTAL

**Comment by zz-jason** [comment]
@breeswish since pr #4347 merged, function `newBaseBuiltinFuncWithTp` does not return error any more, plz update code and fix ci

**Pull request merged**